### ISSUE TITLE
Incorrect `default` type should not be allowed

### ISSUE THREAD

**Issue by sindresorhus** [yargs/yargs-parser#143]
The parser should not allow setting the incorrect type for `default`.

```js
yargs([], {number: ['foo'], default: {foo: 'x'}});
//=> { _: [], foo: 'x' }
```

The above should not be allowed as we defined `foo` as a number, but the `default` is set to a string. This should throw an error to prevent mistakes.

**Comment by bcoe** [comment]
@sindresorhus all of the validation currently happens inside of yargs (https://github.com/yargs/yargs/blob/master/lib/validation.js). Challenges to implementing this feature in `yargs-parser`, from yargs' point of view are:

1. would need to figure out a way to handle translations, even though all the language files are in yargs.
2. would probably need to refactor the `validation.js` in yargs, moving some of the logic towards `yargs-parser.

I think this could be a really interesting task for someone.

**Comment by bcoe** [comment]
refreshing myself with how we handle other strings, we do already have support for translations in yargs-parser, which is good:

```js
  var __ = opts.__ || function (str) {
    return util.format.apply(util, Array.prototype.slice.call(arguments))
  }
```

**Comment by sindresorhus** [comment]
> would need to figure out a way to handle translations, even though all the language files are in yargs.

Give the errors a unique `.code` property code and then detect that in Yargs. `yargs-parser` should not do translations.

**Comment by bcoe** [comment]
@sindresorhus I was about to reply; since this is a development time error, not a user-land error, I wonder fi we could get away with a throw in this one case, even though we're not throwing for any other errors currently in the parser.

**Comment by sindresorhus** [comment]
Yes, throwing would be the correct thing to do here. It's a developer-error, not a end-user error. Throw early.

**Comment by bcoe** [comment]
@sindresorhus, @juergba has been starting to add some more early validation checks in `yargs-parser`, this seems like an ideal candidate for one.

**Comment by bcoe** [comment]
see: https://github.com/yargs/yargs-parser/issues/301

**Comment by Eomm** [comment]
If it doesn't cost too much effort, would you mind to keep an option to coerce the values as it was in previous versions of the module?

An example use case is to save the user's default in the `env` and then use `process.env` to define the `yargs-parser` defaults.
